### PR TITLE
Parallel resolve plugin entries

### PR DESCRIPTION
**Pull request by DoroNahari**
Signed-off-by: Doron Nahari <doron.nahari@sap.com>

#### What it does

Our theia extension is loaded with high number of plugins causes the ide to be started but the plugins are not resolved/ready yet for a few minutes.

This pr resolve the plugins in parallel with regards to their dependencies. i.e. resolve all plugins and collect their dependencies and then resolve all the collected dependencies and so on.
 
Perfomance improvement:
By deploying java-pack vscode extension
Before:
perf-before
After:
perf-after

#### How to test

Add many plugins to Theia at once (preferred with dependencies) and see how they (and their dependencies) resolve in parallel.

For example: deploy vscode java pack: https://marketplace.visualstudio.com/items?itemName=vscjava.vscode-java-pack and check out the console logs for deployment time.

I created a pr with old code and added same logs for comparing the performance: https://github.com/eclipse-theia/theia/pull/6982
#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/plugin-ext/src/main/node/plugin-deployer-impl.ts [hunk 1]
```diff
@@ -130,16 +139,16 @@ export class PluginDeployerImpl implements PluginDeployer {
...
                     }
                 } catch (e) {
                     console.error(`Failed to resolve plugins from '${current}'`, e);
                 }
-            }
-            for (const dependencies of chunk) {
-                for (const [dependency, deployableDependency] of dependencies) {
+            }));
+            for (const dependencies of dependenciesChunk) {
+                for (const [dependency, deployableDependency] of dependencies!) {
```
Comment (amiramw) [inline packages/plugin-ext/src/main/node/plugin-deployer-impl.ts]: dependencies could be undefined and then you'll get "undefined is not iterable"

Comment (marechal-p) [reply packages/plugin-ext/src/main/node/plugin-deployer-impl.ts]: Only non-undefined values seem to be pushed inside the array, doesn't seem to be any issue here. Typing could be refined earlier maybe.

Comment (amiramw) [reply packages/plugin-ext/src/main/node/plugin-deployer-impl.ts]: Right. So maybe change line 120 typing to be `Array< Map<string, string>>`

Comment (amiramw) [reply packages/plugin-ext/src/main/node/plugin-deployer-impl.ts]: and remove ! here

Comment (DoroNahari) [reply packages/plugin-ext/src/main/node/plugin-deployer-impl.ts]: Done

**Comment by akosyakov** [comment]
For optimizations numbers should be provided before and after to see there it really makes the difference. Anyone should be able to follow steps in how to test and reproduce these numbers.

**Comment by DoroNahari** [comment]
> For optimizations numbers should be provided before and after to see there it really makes the difference. Anyone should be able to follow steps in how to test and reproduce these numbers.

Please see the updated `how to test` details.

**Comment by DoroNahari** [comment]
@akosyakov Can it be merged? I'll be glad if it will be included in the next release (0.15.0)

File: packages/plugin-ext/src/main/node/plugin-deployer-impl.ts [hunk 2]
```diff
@@ -100,27 +104,36 @@ export class PluginDeployerImpl implements PluginDeployer {
...
             while (queue.length) {
                 const current = queue.shift()!;
                 if (visited.has(current)) {
                     continue;
+                } else {
+                    workload.push(current);
                 }
                 visited.add(current);
+            }
+            queue = [];
```
Comment (akosyakov) [inline packages/plugin-ext/src/main/node/plugin-deployer-impl.ts]: why is it necessary? it can be reached only if length of queue is 0

**Comment by akosyakov** [comment]
@amiramw @benoitf fine to merge?

**Comment by benoitf** [comment]
@akosyakov yes 👍

**Pull request merged**